<div align="center"><span style="font-family: Arial; color:#0000FF"><b>
    <span style="font-size: x-large">Métodos Numéricos II</span>
    <br>
    <span style="font-size: large">Segundo de Grado en Matemáticas - Curso 2022/23</span>
    <br>
    <span style="font-size: medium">Facultad de Ciencias de la Universidad de Málaga</span>
    <br>
    <span style="font-size: small">Dpto. de Análisis Matemático, Estadística e Investigación Operativa, y Matemática Aplicada</span>
    <br>
    <span style="font-size: small">Profs. Manuel J. Castro y Francisco J. Palma (Área Conocimiento de Matemática Aplicada)</span>
    <br>
    <span style="font-size: medium; color:#FF0000">Práctica número 10</span>
    </b></span></div>

In [2]:
from algoritmos import *

<div align="justify"><span style="font-family: Arial,Helvetica,sans-serif; font-size: medium; color:#000000">
    Se considera el sistema de ecuaciones lineales $A\,X=B$, donde 
\[
A = \left( \begin{array}{rrrrr}
-33 & 36 & -27 & 18 & -9\\
-33 & 51 & -45 & 30 & -15\\
-19 & 38 & -45 & 34 & -17\\
-9 & 18 & -27 & 27 & -15\\
-3 & 6 & -9 & 12 & -9
\end{array} \right) \quad X = \left( \begin{array}{c}
x_1 \\ x_2 \\ x_3 \\ x_4 \\ x_5
\end{array} \right) \quad B = \left( \begin{array}{c}
120 \\ 186\\ 162\\ 99 \\ 36
\end{array} \right)
\]
    </span></div>

<div align="left"><span style="font-family: Arial; color:#000000; font-size: medium">
    <span style="color:#FF0000"><b>Ejercicio 1.</b></span> Al aplicar el método de Gauss con estrategia de pivot parcial al sistema anterior, se obtiene un sistema triangular superior equivalente, que posteriormente se resuelve mediante un proceso de remonte. Escribir dicho sistema triangular superior.
    </span></div>

In [8]:
A = array([[-33, 36, -27, 18, -9], [-33, 51,-45, 30, -15], [ -19, 38, -45, 34, -17], [ -9, 18, -27, 27, -15],
           [ -3, 6, -9, 12, -9]])
print("A = ", A)

B=array([[120],[186],[162], [99],[36]])

#Gauss con pivote parcial pero muestra MA, MB
#MA = GAUSSA
#MB = GAUSSB
def mostrarGauss(A, B): #con pivote parcial
    m, n = shape(A)
    p, q = shape(B)
    if m != n or n != p or q < 1:
        return False, "Error gauss_pp: error en las dimensiones."
    if A.dtype == complex or B.dtype == complex:
        gaussA = array(A, dtype=complex) # a_1, a_2, a_3 ...
        gaussB = array(B, dtype=complex) # b_1, b_2, b_3 ...
    else:
        gaussA = array(A, dtype=float)
        gaussB = array(B, dtype=float)
        
    for k in range(n-1): #va de 0 a n-2  //Aquí empieza el buble principal del método de Gauss => busquemos el pívot
        
        print("\n\nIteración: k =",k+1)
        
        pos = argmax(abs(gaussA[k:, k])) #Elección del pívot (argmax = índice donde se encuentra el máximo de los elementos)
        ik = pos+k                       #Ya que necesitamos el a_kk como la fila 0 (argmax empieza a contar desde 0
        print("Posición del pivote: ik =",ik+1)
        
        if ik != k:     #Si el pivot no está en la posición 0, permutamos las filas (intercambiamos las fila ik-esima y la k-esima)                
            gaussA[[ik, k], :] = gaussA[[k, ik], :]
            gaussB[[ik, k], :] = gaussB[[k, ik], :]
        if abs(gaussA[k, k]) >= 1e-200:
            for i in range(k+1, n):
                gaussA[i, k] = gaussA[i, k]/gaussA[k, k]   #guarda los cocientes donde tenemos los ceros (el programa los ignora
                                                           #pensando que hay ceros, pero nosotros aprovechamos para guardas informació ahí)
                gaussA[i, k+1:] -= gaussA[i, k]*gaussA[k, k+1:]
                gaussB[i, :] -= gaussA[i, k]*gaussB[k, :]
#            print("\n Matriz A_k+1:\n",gaussA)
#            print("\n Matriz B_k+1:\n",gaussB)
                
#    print("\nMatriz triangualar MA:\n",triu(gaussA)) #Imprime la parte diagonal superior de la matriz (en el resto habíamos guardado
                                                      #información economizando, por lo que si la mostrásemos entera, no sería ceros)
#    print("\nSegundo miembro MB:\n",gaussB)
#
    MA=triu(gaussA)
    MB=gaussB
    exito, X = remonte(gaussA, gaussB)
    return exito, X, MA,MB


#######################

exito, X, MA, MB = mostrarGauss(A,B)
if(not exito):
    print("\n Error: no se pudo resolver el sistema")
else:
    print("\n Matriz triangular superior es MA =\n", MA)
    print("\n Vector MB=\n",MB)
    print("\n La solución tras el remonte es X=\n",X)


A =  [[-33  36 -27  18  -9]
 [-33  51 -45  30 -15]
 [-19  38 -45  34 -17]
 [ -9  18 -27  27 -15]
 [ -3   6  -9  12  -9]]


Iteración: k = 1
Posición del pivote: ik = 1


Iteración: k = 2
Posición del pivote: ik = 3


Iteración: k = 3
Posición del pivote: ik = 3


Iteración: k = 4
Posición del pivote: ik = 4

 Matriz triangular superior es MA =
 [[-33.          36.         -27.          18.          -9.        ]
 [  0.          17.27272727 -29.45454545  23.63636364 -11.81818182]
 [  0.           0.           7.57894737  -8.52631579   4.26315789]
 [  0.           0.           0.           4.5         -3.75      ]
 [  0.           0.           0.           0.          -1.5       ]]

 Vector MB=
 [[120.        ]
 [ 92.90909091]
 [-14.68421053]
 [ 11.25      ]
 [ -4.5       ]]

 La solución tras el remonte es X=
 [[1.]
 [4.]
 [2.]
 [5.]
 [3.]]


<div align="left"><span style="font-family: Arial; color:#000000; font-size: medium">
    <span style="color:#FF0000"><b>Ejercicio 2.</b></span> Escribir la solución de dicho sistema, resuelto mediante el método de Gauss con estrategia de pívot parcial.
    </span></div>

In [ ]:
#YA HECHO EN EJERCICIO 1

<div align="left"><span style="font-family: Arial; color:#000000; font-size: medium">
    <span style="color:#FF0000"><b>Ejercicio 3.</b></span> Obtener toda la información que sea posible sobre los autovalores y autovectores de la matriz A, utilizando única y exclusivamente el método de la potencia, el método de la potencia inversa, y el método de la potencia desplazada inversa con desplazamiento de $9$, en los tres casos tomando como vector inicial el primer vector de la base canónica.
    <br>
    Indicar también la localización de los autovalores que aún faltan por calcular. Se valorará positivamente la corrección de los razonamientos a partir de la información disponible.
    </span></div>

In [32]:
print("METODO DE LA POTENCIA\n")
X0 = array([[1],[0],[0],[0],[0]])
exito, norma, valores, vector = potencia(A,X0,inf,200,1e-10)
if(exito):
    print("Norma:",norma)
    print("Valores obtenidos:",valores)
    print("Vector:\n",vector)
    
# Con todo ello, podemos decir que el autovalor dominante obtenido por el método de la potencia
# es -15 (su norma es 15, radio espectral)
#Los autovalores están en el interior de la bola de centro 0 y radio 15


print("\n\nMETODO DE LA POTENCIA INVERSA\n")
X0 = array([[1],[0],[0],[0],[0]])
exito, norma, valores, vector = potenciainv(A,X0,inf,200,1e-10)
if(exito):
    print("Norma:",norma)
    print("Valores obtenidos:",valores)
    print("Vector:\n",vector)
    
#Como el radio espectral de la matriz inversa de A es 0.33333, vemos que su autovalor más pequeño es -0.33333,
#Luego el autovalor más pequeño (en norma) obtenido por el método de la potencia inversa es es -3
#Los autovalore están fuera de la bola de centro 0 y radio 3


print("\n\nMETODO DE LA POTENCIA DESPLAZADA INVERSA\n")
X0 = array([[1],[0],[0],[0],[0]])
exito, norma, valores, vector = potenciadesinv(A,X0,9,inf,200,1e-10)
if(exito):
    print("Norma:",norma)
    print("Valores obtenidos:",valores)
    print("Vector:\n",vector)
    
#En este caso no podemos decir nada ya que nos sale -0.3 y 0.3. Esto nos dice que no hay convergecia 
# Ademas por ejemplo tmb obtenemos 1 y en el vector aparece un -0.75 en 
#esa posición, cuando debería salir un número cercano a cero


#Podemos probar con un desplazamiento de 8.5 por ejemplo

print("\n\nMETODO DE LA POTENCIA DESPLAZADA INVERSA\n")
X0 = array([[1],[0],[0],[0],[0]])
exito, norma, valores, vector = potenciadesinv(A,X0,8.5,inf,200,1e-13)
if(exito):
    print("Norma:",norma)
    print("Valores obtenidos:",valores)
    print("Vector:\n",vector)

#Aquí obtenemos 0.4 => (0.4^-1)+8.5=6 luego 6 es un autovalor


#TRUCO: no nos podemos fijar en los datos del vector que se van a cero. Por tanto, los valores (dentro de la variable "valores")
# que no coincidan con la norma, tienen que tener en la posicion correspondiente del vector obtenido, un número muy cercano a cero

#Si nos preguntasen cual es el autovalor más grande (positivo) dps de 15
#hacemos metodo potencia desplazada inversa -15, puesto que así se pondría el centro en -15, y este método nos daría el autovalor
#más cercano a cero de la desplazada, que luego al convertirla de nuevo, será el más alejado del cero (ver gráficamente)
    

METODO DE LA POTENCIA


Iteración: k =  1
Norma: ||A*X_k|| =  33.0

Iteración: k =  2
Norma: ||A*X_k|| =  8.454545454545459

Iteración: k =  3
Norma: ||A*X_k|| =  24.290322580645167

Iteración: k =  4
Norma: ||A*X_k|| =  10.410358565737049

Iteración: k =  5
Norma: ||A*X_k|| =  20.290470723306584

Iteración: k =  6
Norma: ||A*X_k|| =  11.871159395688322

Iteración: k =  7
Norma: ||A*X_k|| =  18.16279867873538

Iteración: k =  8
Norma: ||A*X_k|| =  12.910366964026316

Iteración: k =  9
Norma: ||A*X_k|| =  16.94228378647604

Iteración: k =  10
Norma: ||A*X_k|| =  13.624305569930472

Iteración: k =  11
Norma: ||A*X_k|| =  16.21168253868798

Iteración: k =  12
Norma: ||A*X_k|| =  14.103104170121957

Iteración: k =  13
Norma: ||A*X_k|| =  15.763147589970874

Iteración: k =  14
Norma: ||A*X_k|| =  14.419039184440752

Iteración: k =  15
Norma: ||A*X_k|| =  15.483494752842743

Iteración: k =  16
Norma: ||A*X_k|| =  14.625282461955326

Iteración: k =  17
Norma: ||A*X_k|| =  15.307454605969712



<div align="left"><span style="font-family: Arial; color:#000000; font-size: medium">
    <span style="color:#FF0000"><b>Ejercicio 4.</b></span> Construir una matriz de orden $𝑛=5$, pentadiagonal, cuyos elementos diagonales valgan 4, y los elementos de las dos diagonales superiores e inferiores $−1$. Tomando como vector inicial aquel que tiene todas sus componentes igual a 1, y aplicando el método de la potencia, calcular su autovalor de módulo mayor y menor; calcular también cond2(𝐴). Ídem para $𝑛=20$.
    </span></div>

In [37]:
#A es una matriz simétrica => autovalores reales

n=5
A=4*eye(n) - eye(n,k=1) -eye(n,k=-1) -eye(n,k=2) -eye(n,k=-2)
print("A=\n",A)
X0=ones((n,1))
print("X0=\n",X0)

print("\n\nMETODO DE LA POTENCIA PARA AUTOVALOR DE MAYOR MODULO")
exito,norma,lambdas,vector = potencia(A,X0,inf,200,1e-10)
if(exito):
    print("\nConvergencia de las normas: ",norma)
    print("\nConvergencia de los cocientes: ", lambdas)
    print("\nConvergencia de los vectores:\n",vector)
else:
    print(norma) #para q imprima un error
    
    
#Nos sale que el autovalor de mayor modulo es 5.47
    

print("\n\nMETODO DE LA POTENCIA INVERSA PARA AUTOVALOR DE MENOR MODULO")
exito,norma,lambdas,vector = potenciainv(A,X0,inf,200,1e-10)
if(exito):
    print("\nConvergencia de las normas: ",norma)
    print("\nConvergencia de los cocientes: ", lambdas)
    print("\nConvergencia de los vectores:\n",vector)
else:
    print(norma) #para q imprima un error
    
    
#Nos sale que el autovalor de menor módulo es 1/0.93 = 1.06
    
print("\nCon_2(A) = ",condicionamiento(A,2))
#Si la matriz fuera singular, no se podría hallar la norma de la inversa pq no existe, se tendría que hacer con valores singulares

A=
 [[ 4. -1. -1.  0.  0.]
 [-1.  4. -1. -1.  0.]
 [-1. -1.  4. -1. -1.]
 [ 0. -1. -1.  4. -1.]
 [ 0.  0. -1. -1.  4.]]
X0=
 [[1.]
 [1.]
 [1.]
 [1.]
 [1.]]


METODO DE LA POTENCIA PARA AUTOVALOR DE MAYOR MODULO

Iteración: k =  1
Norma: ||A*X_k|| =  2.0

Iteración: k =  2
Norma: ||A*X_k|| =  3.5

Iteración: k =  3
Norma: ||A*X_k|| =  5.7142857142857135

Iteración: k =  4
Norma: ||A*X_k|| =  5.75

Iteración: k =  5
Norma: ||A*X_k|| =  5.591304347826087

Iteración: k =  6
Norma: ||A*X_k|| =  5.548989113530326

Iteración: k =  7
Norma: ||A*X_k|| =  5.53054932735426

Iteración: k =  8
Norma: ||A*X_k|| =  5.518724978462474

Iteración: k =  9
Norma: ||A*X_k|| =  5.50981166380474

Iteración: k =  10
Norma: ||A*X_k|| =  5.502758223004412

Iteración: k =  11
Norma: ||A*X_k|| =  5.497096875116414

Iteración: k =  12
Norma: ||A*X_k|| =  5.492529879627507

Iteración: k =  13
Norma: ||A*X_k|| =  5.488836406811336

Iteración: k =  14
Norma: ||A*X_k|| =  5.485844412579323

Iteración: k =  15
Norma: |

<div align="left"><span style="font-family: Arial; color:#000000; font-size: medium">
    <span style="color:#FF0000"><b>Ejercicio 5.</b></span> Dadas las matrices
\[
A = \left( \begin{array}{rrrr}
9 & 1 & -2 & 1 \\ 1 & 8 & -3 & 2 \\ -2 & -3 & 7 & -1 \\ 1 & -2 & -1 & 6
\end{array} \right) \quad A = \left( \begin{array}{rrrr}
9 & 1 & -2 & 1 \\ 1 & 8 & -3 & -2 \\ -2 & -3 & 7 & -1 \\ 1 & -2 & -1 & 6
\end{array} \right)
\]
\[
A = \left( \begin{array}{rrrr}
1 & 2 & 3 & 4 \\ 2 & 1 & 4 & 3 \\ 3 & 4 & 1 & 2 \\ 4 & 3 & 2 & 1
\end{array} \right)
\]
\[
A = \left( \begin{array}{rrrr}
1 & -1 & 3 & 4 \\ -1 & 4 & 0 & -1 \\ 3 & 0 & 0 & -3 \\ 4 & -1 & -3 & 1
\end{array} \right) \quad A = \left( \begin{array}{rrrr}
0 & 1 & -2 & 3 \\ -1 & 0 & 4 & -5 \\ 2 & -4 & 0 & 6 \\ -3 & 5 & -6 & 0
\end{array} \right)
\]
calcular mediante el método de la potencia y sus variantes toda la información posible sobre sus autovalores y autovectores.
    </span></div>

In [19]:
A = array([[9, 1, -2, 1], [1, 8,-3, 2], [-2, -3, 7, -1], [1, -2, -1, 6]])
print("A = \n", A)
X0 = array([[1],[0],[0],[0]])

print("\n\nMETODO DE LA POTENCIA\n\n")
exito, norma, lambdas, vector = potencia(A,X0,inf,200,1e-10)
if(exito):
    print("Convergencia de la norma:",norma)
    print("Convergencia de los lambdas:",lambdas)
    print("Convergencia de los vectores:\n",vector)
else:
    print(norma)
    
    
print("\n\nMETODO DE LA POTENCIA INVERSA\n\n")
exito, norma, lambdas, vector = potenciainv(A,X0,inf,200,1e-10)
if(exito):
    print("Convergencia de la norma:",norma)
    print("Convergencia de los lambdas:",lambdas)
    print("Convergencia de los vectores:\n",vector)
else:
    print(norma)

    
print("\n\nMETODO DE LA POTENCIA DESPLAZADA\n\n")
exito, norma, lambdas, vector = potenciades(A,X0,3,inf,200,1e-10)
if(exito):
    print("Convergencia de la norma:",norma)
    print("Convergencia de los lambdas:",lambdas)
    print("Convergencia de los vectores:\n",vector)
else:
    print(norma)
    



print("\n\nMETODO DE LA POTENCIA INVERSA DESPLAZADA\n\n")
exito, norma, lambdas, vector = potenciadesinv(A,X0,5.5,inf,200,1e-10)
if(exito):
    print("Convergencia de la norma:",norma)
    print("Convergencia de los lambdas:",lambdas)
    print("Convergencia de los vectores:\n",vector)
else:
    print(norma)
    


A = 
 [[ 9  1 -2  1]
 [ 1  8 -3  2]
 [-2 -3  7 -1]
 [ 1 -2 -1  6]]


METODO DE LA POTENCIA



Iteración: k =  1
Norma: ||A*X_k|| =  9.0

Iteración: k =  2
Norma: ||A*X_k|| =  9.666666666666666

Iteración: k =  3
Norma: ||A*X_k|| =  10.28735632183908

Iteración: k =  4
Norma: ||A*X_k|| =  10.81005586592179

Iteración: k =  5
Norma: ||A*X_k|| =  11.21064599483204

Iteración: k =  6
Norma: ||A*X_k|| =  11.494749361533428

Iteración: k =  7
Norma: ||A*X_k|| =  11.684854682756436

Iteración: k =  8
Norma: ||A*X_k|| =  11.806947017504143

Iteración: k =  9
Norma: ||A*X_k|| =  11.88319786170418

Iteración: k =  10
Norma: ||A*X_k|| =  11.929934452342877

Iteración: k =  11
Norma: ||A*X_k|| =  11.95822201583649

Iteración: k =  12
Norma: ||A*X_k|| =  11.975196620595698

Iteración: k =  13
Norma: ||A*X_k|| =  11.985321567530278

Iteración: k =  14
Norma: ||A*X_k|| =  11.991334737842003

Iteración: k =  15
Norma: ||A*X_k|| =  11.994894426325263

Iteración: k =  16
Norma: ||A*X_k|| =  11.996996473

- Método de la Potencia:

Con esto vemos que 12 es el autovalor de mayor modulo (los demás autovalores están en cincunf centro 0 y radio 15)


- Método de la Potencia Inversa:

Vemos esto vemos que el autovalor de menor modulo es 1/0.2 = 5 (demás autovalores fuera de circunf centro 0 y radio 5)


- Método de la Potencia Desplazada 7: 

Volvemos a obtener 12 
De hecho si metiendo varios números solo he obtenido 12 y 5




- Método de la Potencia Inversa 8:
Se obtiene -1 => (-1)^-1 + 8 = 7 autovalor de A

- Método de la Potencia Inversa 3:
Se obtiene 0.5=> (0.5)^-1 + 3 = 5 autovalor de A (ya lo sabíamos)

- Método de la Potencia Inversa 4.5:
Se obtiene 2=> (2)^-1 + 4.5 = 5 autovalor de A (ya lo sabíamos)



Falta un autovalor de A más
